runScenario() is sometimes using old values of hp and energy

franklist opened the pull request:
Currently the values of hp and energy are getting changed during calls to simulate, so the simulation AI is making decisions based on inflated HP values and old energy values.  This small pull request sets the correct starting values of these parameters during all calls to runScenario, not just during evaluateMatchup().

This fix appears to resolve an unintentional bug.  The values of hp and energy, along with other stats, are reset to their starting values during simulate, as simulate calls each Pokemon's reset().  The starting values are updated for this at the beginning of each call to evaluateMatchup, which is correct, but there are additional calls to runScenario in other function calls like decideSwitch and decideShield, so the value of HP will briefly be reset to the old value for the duration of those calls to simulate().  Shields and buffs could be affected as well, but those starting values tend to stay updated with only evaluateMatchup in my testing.

The issue appears to be isolated to runScenario().  Other parts of the AI call the HP and energy directly without resetting them, and the stale HP value is currently erased at the end of runScenario().

I tried to come up with an easy example. If facing a Master League lead Kyogre against a team of Melmetal, Yveltal, and Togekiss, after the first matchup, assuming no switch occurs, Kyogre probably has a little HP left.  If the AI overestimates Kyogre's HP and energy then the AI may, depending on the scenario, prefer switching in Yveltal in instead of Togekiss, even though Togekiss is better at farming down a low-HP Pokemon, and is a better switch anyways.

I used a custom AI vs AI simulation to check that this actually improves the AI. The updated AI beats the traditional AI roughly 54% of the time, using a sample of common Great League teams, so this is a minor but measureable improvement to the AI.

There were a few details I wasn't sure of like whether to use "pokemon.startHp =" or "pokemon.setStartingHp()" and whether it's overkill to update shields and buffs as well since they stay pretty in sync in the old version.  Let me know if there are any issues.

pvpoke reviewed (approved):
Thanks so much for finding and fixing this issue! I'll get this merged and uploaded.

pvpoke commented:
Oops! I had someone report that the bot is able to use 3 shields now. I've reverted the changes for the moment but posting here in case you get a chance to look into it!

https://twitter.com/PogoGastly/status/1417955241664909313

franklist commented:
I was able to confirm the extra shield issue, sorry that I didn't catch that during testing.

It seems to come from the fact that in the previous and current code runscenario saves and then restore startingShields instead of shields, which in my version was usually still 2 at the start of the first call to evaluateScenario shielding.  I'm not sure of the reasoning for that, as there's no guarantee, especially with the code I cleaned out from evaluateMatchup, that resetting to the starting stats will reflect the values at the beginning of the scenario.  This leaves three choices:

1. If you think the saved start stats in runScenario shouldn't be using startingShields in the first place, change the start stats to use (current) shields instead of startingShields
2. If you just want startingShields to reflect current values, move the lines I added above the start stats so that starting values reflect the current values.
3. If you want behavior most like the old code, put the lines back in evaluateMatchup to sync starting shields and current shields, even though similar code will just be called a few lines later in runScenario.

Which do you think is the best solution?

pvpoke commented:
No worries, it happens! Number 2 seems like the most straightforward solution and on my end it seems to resolve the issue. Would you prefer to submit this change as another PR or should I just commit it on my side?
